### PR TITLE
Feat/td 32275

### PR DESCRIPTION
**Pull request by menshibin**

### CODE REVIEW

File: src/main/java/com/taosdata/flink/common/TDengineConfigParams.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+    public static final String VALUE_DESERIALIZER = "value.deserializer";
+
+    public static final String TD_SOURCE_TYPE = "td.source.type";
+
+    public static final String TD_BATCH_MODE = "td.batch.mode";
+
+    public static final String TD_JDBC_URL = "url";
+
+    public static final String TD_DATABASE_NAME = "database_name";
```
Comment (sheyanjie-qq) [inline src/main/java/com/taosdata/flink/common/TDengineConfigParams.java:15]: database.name for consistency?

File: src/main/java/com/taosdata/flink/example/Main.java [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+import java.util.Arrays;
+import java.util.List;
+import java.util.Properties;
+
+import org.apache.flink.streaming.api.CheckpointingMode;
+
+
+
+public class Main {
+    static String jdbcUrl = "jdbc:TAOS-WS://192.168.1.98:6041?user=root&password=taosdata";
```
Comment (sheyanjie-qq) [inline src/main/java/com/taosdata/flink/example/Main.java:41]: use localhost to run ci

File: src/main/java/com/taosdata/flink/sink/entity/SinkErrorNumbers.java [hunk 3]
```diff
@@ -10,13 +10,19 @@ public class SinkErrorNumbers {
 
     public static final int ERROR_SQL_EXECUTION_NO_RESULTS = 0x1002;
```
Comment (sheyanjie-qq) [inline src/main/java/com/taosdata/flink/sink/entity/SinkErrorNumbers.java:12]: please use error number 0x3000+

File: src/main/java/com/taosdata/flink/sink/TDengineSqlWriter.java [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+            StringBuilder sb = new StringBuilder("(");
+            for (int i = 0; i < this.sinkMetaInfos.size(); i++) {
+                sb.append(getStringParam(sinkRecord.getColumnParams().get(i), sinkMetaInfos.get(i).getFieldType().getTypeNo()));
+                if (i < sinkMetaInfos.size() - 1) {
+                    sb.append(",");
+                }
+            }
+            sb.append(")") ;
+            // Splicing SQL, writing nearly 1M to TDengine
+            if ((executeSqls.length() + sb.length()) >= ONE_MILLION) {
```
Comment (sheyanjie-qq) [inline src/main/java/com/taosdata/flink/sink/TDengineSqlWriter.java:158]: to make sure avoid sql too long,  we need check every record before it added to sql.

File: src/main/java/com/taosdata/flink/sink/TDengineStmtWriter.java [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+            throw new RuntimeException(e);
+        }
+
+        LOG.info("connect websocket url ok");
+    }
+
+    @Override
+    public void write(IN element, Context context) throws IOException, InterruptedException {
+        List<TDengineSinkRecord> records = serializer.serialize(element, sinkMetaInfos);
+        System.out.println("sink write start！");
```
Comment (sheyanjie-qq) [inline src/main/java/com/taosdata/flink/sink/TDengineStmtWriter.java:102]: remove all system out

File: src/main/java/com/taosdata/flink/sink/TDengineStmtWriter.java [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+        if (records == null || records.isEmpty()) {
+            LOG.warn("element serializer result is null!");
+            return;
+        }
+
+        try {
+            lock.lock();
```
Comment (sheyanjie-qq) [inline src/main/java/com/taosdata/flink/sink/TDengineStmtWriter.java:109]: why need a lock?

File: src/main/java/com/taosdata/flink/sink/TDengineStmtWriter.java [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+            for (TDengineSinkRecord record : records) {
+                setStmtParam(pstmt, record.getColumnParams());
+                pstmt.addBatch();
+                recodeCount.incrementAndGet();
+            }
+            if (recodeCount.get() >= batchSize) {
```
Comment (sheyanjie-qq) [inline src/main/java/com/taosdata/flink/sink/TDengineStmtWriter.java:115]: check the batchsize in loop

File: src/main/java/com/taosdata/flink/source/entity/SourceErrorNumbers.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
+package com.taosdata.flink.source.entity;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public class SourceErrorNumbers {
+
+    public static final int ERROR_SERVER_ADDRESS = 0x2000;
```
Comment (sheyanjie-qq) [inline src/main/java/com/taosdata/flink/source/entity/SourceErrorNumbers.java:8]: jdbc use 0x2000+ error  numbers

File: src/main/java/com/taosdata/flink/source/enumerator/TDengineSourceEnumerator.java [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+     */
+    private Deque<String> splitByTimestamp() {
+        Deque<String> unassignedSql = new ArrayDeque<>();
+        TimestampSplitInfo timestampSplitInfo = sourceSql.getTimestampSplitInfo();
+        if (timestampSplitInfo != null && !Strings.isNullOrEmpty(timestampSplitInfo.getFieldName()) && timestampSplitInfo.getEndTime() > timestampSplitInfo.getStartTime()) {
+            long timeDifference = timestampSplitInfo.getEndTime() - timestampSplitInfo.getStartTime();
+            long nCount = 0;
+            // Obtain the number of separable tasks
+            if (timestampSplitInfo.getInterval() > 0) {
+                nCount = timeDifference / timestampSplitInfo.getInterval();
```
Comment (sheyanjie-qq) [inline src/main/java/com/taosdata/flink/source/enumerator/TDengineSourceEnumerator.java:148]: how it wil be if timeDifference % timestampSplitInfo.getInterval() is not zero

File: src/main/java/com/taosdata/flink/cdc/serializable/TDengineCdcSplitSerializer.java [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+            out.flush();
+            return baos.toByteArray();
+        }
+    }
+
+    @Override
+    public TDengineCdcSplit deserialize(int version, byte[] serialized) throws IOException {
+        try (ByteArrayInputStream bais = new ByteArrayInputStream(serialized);
+             DataInputStream in = new DataInputStream(bais)) {
+            String splitId = in.readUTF();
```
[the remark this answers is not in the document]
Comment (menshibin) [reply src/main/java/com/taosdata/flink/cdc/serializable/TDengineCdcSplitSerializer.java:47]: 这个字段目前没用

File: src/main/java/com/taosdata/flink/cdc/TDengineCdcSource.java [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+    private boolean isBatchMode = false;
+
+    private boolean isAutoCommit = false;
+
+    private final Class<OUT> typeClass;
+    public TDengineCdcSource(String topic, Properties properties, Class<OUT> typeClass) {
+        this.topic = topic;
+        this.properties = properties;
+        this.typeClass = typeClass;
+        String batchMode = this.properties.getProperty(TDengineCdcParams.TMQ_BATCH_MODE, "false");
```
[the remark this answers is not in the document]
Comment (menshibin) [reply src/main/java/com/taosdata/flink/cdc/TDengineCdcSource.java:46]: 批量不是常规用法，table sql 是不支持的

File: src/main/java/com/taosdata/flink/sink/TDengineStmtWriter.java (hunk 6 above)
Comment (menshibin) [reply src/main/java/com/taosdata/flink/sink/TDengineStmtWriter.java:108]: 多线程写入

File: src/main/java/com/taosdata/flink/sink/TDengineSqlWriter.java (hunk 4 above)
Comment (menshibin) [reply src/main/java/com/taosdata/flink/sink/TDengineSqlWriter.java:158]: 上面循环是一个sql的长度，不能放到上面

File: README.md [hunk 12]
```diff
@@ -1,2 +1,626 @@
...
+
+With the help of TDengine's Flink connector, Apache Flink can seamlessly integrate with the TDengine database. On the one hand, it can accurately store the results obtained after complex calculations and deep analysis into the TDengine database, achieving efficient storage and management of data; On the other hand, it is also possible to quickly and stably read massive amounts of data from the TDengine database, and conduct comprehensive and in-depth analysis and processing on this basis, fully tapping into the potential value of the data, providing strong data support and scientific basis for enterprise decision-making, greatly improving the efficiency and quality of data processing, and enhancing the competitiveness and innovation ability of enterprises in the digital age.
+
+## Preconditions
```
Comment (zitsen) [inline README.md:7]: Use **Prerequisites**

File: README.md [hunk 13]
```diff
@@ -1,2 +1,626 @@
...
+| TDengineConfigParams.PROPERTYKEY_RECONNECT_RETR_COUNT   |                                               integer                                                | number of automatic reconnection retries, default value 3                                                                                                                           | only takes effect when PROPERTYKEY-INABLE AUTO-RECONNECT is true|
+| TDengineConfigParams.PROPERTYKEYDISABLE_SSL_CERTVNet    |                                               boolean                                                | Disable SSL certificate verification. true:  close, false:  Not closed. The default is false                                                                                        ||
+
+#### Use Sink connector
+Write the received RowData type data into TDengine example:
```
Comment (zitsen) [inline README.md:408]: Use the same coding style for markdown, leave a blank line between Header and Detail.

File: README.md [hunk 14]
```diff
@@ -1,2 +1,626 @@
...
+    Thread.sleep(6000L);
+    jobClient.cancel().get();
+    System.out.println("testTDengineCdcToTdSink finish！");
+}
+```
+
+### Table SQL
+
+ETL (Extract, Transform, Load) data processing: Flink SQL with JDBC can be used to extract data from multiple different data source databases (such as TDengine, MySQL, Oracle, etc.), perform transformation operations (such as data cleaning, format conversion, associating data from different tables, etc.) in Flink, and then load the processed results into the target data source (such as TDengine, MySQL, etc.).
+#### Source connector
```
Comment (zitsen) [inline README.md:489]: Same here, add new before header.

File: README.md [hunk 15]
```diff
@@ -1,2 +1,626 @@
...
+| td.jdbc.mode          | strng | connector type, set source, cdc, sink| |
+| table.name            | string | Original or target table name ||
+| scan.query            | string | SQL statement to retrieve data||
+| sink.db.name          | string | Target database name||
+| sink.superstable.name | string | Write the name of the superstable||
+| sink.batch.size       | integer | batch size written||
+| sink.table.name       | string | Name of the regular table or sub table written||
+
+#### Example of using Source connector
+```java
```
Comment (zitsen) [inline README.md:506]: new line.

File: README.md [hunk 16]
```diff
@@ -1,2 +1,626 @@
...
+        tableEnv.executeSql(tdengineSourceTableDDL);
+        tableEnv.executeSql(tdengineSinkTableDDL);
+        tableEnv.executeSql("INSERT INTO sink_meters SELECT ts, `current`, voltage, phase, location, groupid, tbname FROM `meters`");
+    }
+
+
+```
+
+#### CDC connector
+Parameter configuration instructions:
```
Comment (zitsen) [inline README.md:556]: new line.

File: src/main/java/com/taosdata/flink/cdc/serializable/RowDataCdcDeserializer.java [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+     * @param dbName database name
+     * @return RowData
+     * @throws DeserializerException
+     * @throws SQLException
+     */
+    @Override
+    public RowData deserialize(ResultSet data, String topic, String dbName) throws DeserializerException, SQLException {
+        ResultSetMetaData metaData = data.getMetaData();
+        GenericRowData row = new GenericRowData(metaData.getColumnCount());
+        for (int i = 1; i <= metaData.getColumnCount(); i++) {
```
[the remark this answers is not in the document]
Comment (menshibin) [reply src/main/java/com/taosdata/flink/cdc/serializable/RowDataCdcDeserializer.java:31]: ResultSet from 1 begine

**Pull request merged**